src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs

```diff
@@ -153,128 +153,118 @@ public int GetUpperBound(int dimension)
+        public unsafe T GetValue<T>(params int[] indices) where T : unmanaged
+        {
+            if (sizeof(T) != Type.ComponentSize)
+                throw new ArgumentException($"{typeof(T).Name} is 0x{sizeof(T):x} bytes but the array element is 0x{Type.ComponentSize:x}.");
 
-                dataByteOffset += 2 * sizeof(int);
-            }
-            else
-            {
-                if ((uint)valueOffset >= Length)
-                    throw new ArgumentOutOfRangeException(nameof(index));
-            }
+            return ReadValue<T>(indices);
+        }
 
+        public ClrValueType GetStructValue(int index)
+        {
+            if (Type.ComponentType is null)
+                return default;
 
-            int elementSize = isObjectReference ? UIntPtr.Size : Type.ComponentSize;
-            int valueByteOffset = dataByteOffset + valueOffset * elementSize;
+            if (Type.ComponentType.IsObjectReference)
+                throw new InvalidOperationException($"{Type} does not contain value type instances.");
 
-            if (isObjectReference)
-            {
-                return Type.ClrObjectHelpers.DataReader.Read<UIntPtr>(Address + (ulong)valueByteOffset).ToUInt64();
-            }
-            else
-            {
-                return Address + (ulong)valueByteOffset;
-            }
+            ulong address = GetElementAddress(Type.ComponentSize, index);
+            return new ClrValueType(address, Type.ComponentType, interior: true);
         }
 
-        public ClrValueType GetStructValue(int index) => GetStructValue(index, false);
-
+        [Obsolete("Use GetStructValue(int) instead.")]
         public ClrValueType GetStructValue(int index, bool interior)
         {
-            if (Type.ComponentType == null)
-                return new ClrValueType();
+            if (Type.ComponentType is null)
+                return default;
 
             if (Type.ComponentType.IsObjectReference)
                 throw new InvalidOperationException($"{Type} does not contain value type instances.");
 
-            ulong address = GetItemAddress(index, false);
+            ulong address = GetElementAddress(Type.ComponentSize, index);
             return new ClrValueType(address, Type.ComponentType, interior);
         }
 
+        public ClrValueType GetStructValue(params int[] indices)
+        {
+            if (Type.ComponentType is null)
+                return default;
+
+            if (Type.ComponentType.IsObjectReference)
+                throw new InvalidOperationException($"{Type} does not contain value type instances.");
+
+            ulong address = GetElementAddress(Type.ComponentSize, indices);
+            return new ClrValueType(address, Type.ComponentType, interior: true);
+        }
 
         public ClrObject GetObjectValue(int index)
         {
-            if (Type.ComponentType == null)
-                return new ClrObject();
+            if (Type.ComponentType is null)
```

Comment:
Thinking about this more, I think we should remove this null check/return default.  We can successfully get an object value if `ComponentType` is null, we don't use it at all in the size calculation, and the ClrType used for the object comes from reading the method table with ClrHeap.GetObject.

Reply:
What about the reference/value type validation?

```cs
if (Type.ComponentType?.IsObjectReference == false) throw
```

Reply:
Yeah that seems right to me.  I'll wrap that into a change and merge this one as is.